### PR TITLE
[js/webgpu] support using uniform buffer

### PR DESCRIPTION
**Pull request by fs-eire**
### Description
support using uniform buffer.

This PR allows to use uniform buffer in shader program, so that some runtime information (eg. input/output shape) is no longer need to be hardcoded into shader code.

There are 2 commits in this PR:
- [667f31c](https://github.com/microsoft/onnxruntime/pull/17803/commits/667f31c83dcc5324427966063b62e9128fa30558): framework changes to support uniform buffer, as well as updates in program manager, gpu data manager and indices helper.
- [09e1d2a](https://github.com/microsoft/onnxruntime/pull/17803/commits/09e1d2ad1de521f94a2eb27718262d1ccafa4b02): an example change for operator `Transpose` to use input's rank-only instead of dims as shader key. With this change, model mobilenetv2-12 shader compile times dropped from 71 to 52.

### CODE REVIEW

File: js/web/lib/wasm/jsep/webgpu/ops/reduce.ts [hunk 1]
```diff
@@ -31,8 +31,8 @@ export type ReduceOp =
 
 const noOp: ReduceOp = (input) => ['', '', `var value = ${input.getByOffset('inputOffset')};`, ''];
 export const createReduceProgramInfo =
-    (metadata: ProgramMetadata, inputs: readonly TensorView[], reduceOp: ReduceOp, axesInput: number[],
-     outputDataType: DataType, keepDims = false, noopWithEmptyAxes = false): ProgramInfo => {
+    (name: string, shaderCache: ProgramShaderCacheInfo, inputs: readonly TensorView[], reduceOp: ReduceOp,
+     axesInput: number[], outputDataType: DataType, keepDims = false, noopWithEmptyAxes = false): ProgramInfo => {
```
Comment (satyajandhyala) [inline js/web/lib/wasm/jsep/webgpu/ops/reduce.ts:35]: Can axesInput be readonly?

File: js/web/lib/wasm/jsep/webgpu/gpu-data-manager.ts [hunk 2]
```diff
@@ -323,13 +334,19 @@ class GpuDataManagerImpl implements GpuDataManager {
...
+      });
+    });
 
     this.storageCache.forEach((storage) => {
       storage.gpuData.buffer.destroy();
     });
 
     this.storageCache = new Map();
     this.freeBuffers = new Map();
+    this.freeUniformBuffers = new Map();
```
Comment (axinging) [inline js/web/lib/wasm/jsep/webgpu/gpu-data-manager.ts:349]: How about 
```
    this.storageCache = null;
    this.freeBuffers = null;
    this.freeUniformBuffers = null;
```
BTW, gpuDataManager.dispose wasn't used in backedn-webgpu.

File: js/web/lib/wasm/jsep/webgpu/types.ts [hunk 3]
```diff
@@ -26,64 +26,96 @@ export interface TensorInfo {
...
    */
-  variables?: ProgramVariable[];
+  name: string;
+
   /**
-   * tensor info for outputs
+   * gpu data types for each input
    */
-  outputs: TensorInfo[];
+  inputTypes: GpuDataType[];
```
Comment (qjia7) [inline js/web/lib/wasm/jsep/webgpu/types.ts:94]: It seems that the `inputTypes` is not very useful. I only find it's used in https://github.com/microsoft/onnxruntime/blob/main/js/web/lib/wasm/jsep/backend-webgpu.ts#L224. Not sure whether we can remove it.

**Review by guschmue — APPROVED** [review]
lgtm. I ran this with 40+ models - did not see issues there.

**Review by satyajandhyala — APPROVED** [review]

File: js/web/lib/wasm/jsep/webgpu/gpu-data-manager.ts (hunk 2 above)
Comment (fs-eire) [reply js/web/lib/wasm/jsep/webgpu/gpu-data-manager.ts:349]: This is good suggestion. However I tried to write it like this, and typescript reports:
> Type 'null' is not assignable to type 'Map<number, StorageCacheValue>

I don't know if there is a better way to do this. If I define `storageCache` as nullable type, every time when I use it I need to do a null check, which is not necessary. Not sure what is the best practice for this typescript scenario.

And I guess this is why assigning to new Map was there at the first place.

**Pull request merged**